### PR TITLE
Add support for multi-line field options

### PR DESCRIPTION
**Pull request by wouterj**
Refs https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#field-lists

### CODE REVIEW

**Comment by greg0ire** [comment]
Out of curiosity, why do you want to implement this? I suppose you want to improve Symfony's or Doctrine's docs, maybe a specific page?

File: lib/Parser/Directive.php [hunk 1]
```diff
@@ -57,6 +57,6 @@ public function getOptions(): array
      */
     public function setOption(string $key, $value): void
     {
-        $this->options[$key] = $value;
+        $this->options[$key] = $value === '' ? true : $value;
```
Comment (greg0ire) [inline lib/Parser/Directive.php]: How does this relate to your PR?

File: lib/Parser/DocumentParser.php [hunk 2]
```diff
@@ -109,6 +109,9 @@ final class DocumentParser
     /** @var string|null */
     private $listMarker = null;
 
+    /** @var array{name: string,body: string,offset: int}|null */
```
Comment (greg0ire) [inline lib/Parser/DocumentParser.php]: We usually add a space after commas in such situations.

```suggestion
    /** @var array{name: string, body: string, offset: int}|null */
```

File: lib/Parser/LineDataParser.php [hunk 3]
```diff
@@ -233,4 +233,20 @@ public function parseDefinitionList(array $lines): DefinitionList
 
         return new DefinitionList($definitionList);
     }
+
+    /**
+     * @return array{offset: int, name: string, body: string}
```
Comment (greg0ire) [inline lib/Parser/LineDataParser.php]:
```suggestion
     * @return array{offset: int, name: string, body: string}|null
```

File: tests/Parser/LineCheckerTest.php [hunk 4]
```diff
@@ -91,4 +91,14 @@ public function testIsDefinitionListEnded(): void
...
         self::assertFalse($this->lineChecker->isDefinitionListEnded('', '    Definition'));
     }
+
+    public function testIsFieldList(): void
+    {
+        self::assertTrue($this->lineChecker->isFieldList(':glob:'));
+        self::assertTrue($this->lineChecker->isFieldList(':Date: 2001-08-16'));
+        self::assertTrue($this->lineChecker->isFieldList('    :Date: 2001-08-16'));
+        self::assertTrue($this->lineChecker->isFieldList(':Date published: 2001-08-16'));
+        self::assertTrue($this->lineChecker->isFieldList(':Date:published: 2001-08-16'));
```
Comment (greg0ire) [inline tests/Parser/LineCheckerTest.php]: > but colons (":") inside of field names must be backslash-escaped when followed by whitespace

That case does not seem to be tested. Also, the test determines whether there is a field list or not, but it does not say which part is the field name, and which is the body. Shouldn't there be a test for that?

**Comment by wouterj** [comment]
> Out of curiosity, why do you want to implement this? I suppose you want to improve Symfony's or Doctrine's docs, maybe a specific page?

Oh, good question! :facepalm: 

I'm planning to add alt texts to all figures in the Symfony docs (ref https://github.com/symfony/symfony-docs/pull/16895 ). As these can be quite long, I want to use line breaks to conform to our soft 72 character limit in the Symfony docs.

**Comment by wouterj** [comment]
Thanks for the review! I've refactored the changes a bit, to use a VO instead of the array structures. I also added the test cases you suggested.

File: lib/Parser/FieldOption.php [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+    /** @var string */
+    private $body;
+
+    /** @var int */
+    private $lineCount = 0;
+
+    public function __construct(string $name, int $offset, string $body)
+    {
+        $this->name   = str_replace(' \:', ' :', $name);
```
Comment (greg0ire) [inline lib/Parser/FieldOption.php]: >     but colons (":") inside of field names must be backslash-escaped when followed by whitespace

:thinking: it says "followed", is this really what should be done here? Or should `'\: ` be replaced with `: `?

File: tests/Parser/LineDataParserTest.php [hunk 6]
```diff
@@ -52,6 +52,36 @@ public function getTestLinks(): array
...
+            self::assertNull($actual);
+        } else {
+            self::assertNotNull($actual);
+            self::assertSame($name, $actual->getName());
+            self::assertSame($body, $actual->getBody());
+        }
+    }
+
+    /**
+     * @return mixed[]
```
Comment (greg0ire) [inline tests/Parser/LineDataParserTest.php]:
```suggestion
     * @return array{string, string|null, true|string}
```

File: tests/Parser/LineDataParserTest.php [hunk 7]
```diff
@@ -52,6 +52,36 @@ public function getTestLinks(): array
         ];
     }
 
+    /**
+     * @param mixed $body
```
Comment (greg0ire) [inline tests/Parser/LineDataParserTest.php]:
```suggestion
     * @param true|string $body
```

File: lib/Parser/FieldOption.php (hunk 5 above)
Comment (wouterj) [reply lib/Parser/FieldOption.php]: Haha, I finally understand the backslash thing now. Fixed it

**Review by greg0ire — APPROVED** [review]

**Comment by greg0ire** [comment]
Thanks @wouterj !

**Pull request merged**